pkg/ccl/backupccl/restore_checkpointing_test.go

```diff
@@ -0,0 +1,175 @@
+
+	restoreQuery := `RESTORE DATABASE r1 FROM 'nodelocal://0/test-root' WITH detached, new_db_name=data2`
+
+	backupTableID := sqlutils.QueryTableID(t, conn, "r1", "public", "foo")
+
+	var jobID jobspb.JobID
+	do := func(query string, check inProgressChecker) {
+		t.Logf("checking query %q", query)
+
+		var totalExpectedResponses int
+		if strings.Contains(query, "RESTORE") {
+			// We expect restore to process each file in the backup individually.
+			// SST files are written per-range in the backup. So we expect the
+			// restore to process #(ranges) that made up the original table.
+			totalExpectedResponses = 5
+		} else {
+			t.Fatal("expected query to be either a backup or restore")
+		}
+		jobDone := make(chan error)
+		allowResponse = make(chan struct{}, totalExpectedResponses)
+
+		go func() {
+			_, err := conn.Exec(query)
+			jobDone <- err
+		}()
+
+		// Allow half the total expected responses to proceed.
+		for i := 0; i < totalExpectedResponses/2; i++ {
+			allowResponse <- struct{}{}
+		}
+
+		err := retry.ForDuration(testutils.DefaultSucceedsSoonDuration, func() error {
+			return check(ctx, inProgressState{
+				DB:            conn,
+				backupTableID: backupTableID,
+				dir:           dir,
+				name:          "foo",
+			})
+		})
+
+		// Close the channel to allow all remaining responses to proceed. We do this
+		// even if the above retry.ForDuration failed, otherwise the test will hang
+		// forever.
+		close(allowResponse)
+
+		if err := <-jobDone; err != nil {
+			t.Fatalf("%q: %+v", query, err)
+		}
+
+		if err != nil {
+			t.Log(err)
+		}
+	}
+
+	progressQuery := `select crdb_internal.pb_to_json('cockroach.sql.jobs.jobspb.Progress', progress) as progress from system.jobs where id=$1`
+
+	var progressMessage string
+	//var unused interface{}
+
+	checkFraction := func(ctx context.Context, ip inProgressState) error {
+		jobID, err := ip.latestJobID()
+		if err != nil {
+			return err
+		}
+		var fractionCompleted float32
+		if err := ip.QueryRow(
+			`SELECT fraction_completed FROM crdb_internal.jobs WHERE job_id = $1`,
+			jobID,
+		).Scan(&fractionCompleted); err != nil {
+			return err
+		}
+		ip.QueryRow(progressQuery, jobID).Scan(&progressMessage)
+		t.Logf(progressMessage)
+		if fractionCompleted < 0.01 || fractionCompleted > 0.99 {
+			return errors.Errorf(
+				"expected progress to be in range [0.01, 0.99] but got %f",
+				fractionCompleted,
+			)
+		}
+		return nil
+	}
+
+	do(restoreQuery, checkFraction)
+
+	sqlDB.QueryRow(t, `SELECT job_id FROM crdb_internal.jobs ORDER BY created DESC LIMIT 1`).Scan(&jobID)
+	sqlDB.QueryRow(t, progressQuery, jobID).Scan(&progressMessage)
+	require.NotNil(t, progressMessage)
+	require.Contains(t, progressMessage, "\"wallTime\": \"1\"")
+	jobProgress := jobutils.GetJobProgress(t, sqlDB, jobID)
+	require.NotNil(t, jobProgress)
+
+	sqlDB.Exec(t, `PAUSE JOB $1`, jobID)
+	jobutils.WaitForJobToPause(t, sqlDB, jobID)
+	sqlDB.Exec(t, `RESUME JOB $1`, jobID)
```

Comment:
A few notes:
- could you explain in words what you are testing after the job resumes? 
- We spoke about  verifying that all data was restored, could you add a check for that?
- Where does this hardcoded number 7 come from? Why should the frontier length be equal to 7?